master/internal/stream/system.go

```diff
@@ -36,41 +43,77 @@ type SubscriptionSet struct {
 	// Experiments *subscriptionState[*ExperimentMsg, ExperimentSubscriptionSpec]
 }
 
-// subscriptionState contains per-type subscription state
+// subscriptionState contains per-type subscription state.
 type subscriptionState[T stream.Msg, S any] struct {
-	Subscription stream.Subscription[T]
-	FilterMaker FilterMaker[T, S]
-	CollectStartupMsgs CollectStartupMsgsFunc[S]
+	Subscription               stream.Subscription[T]
+	FilterMaker                FilterMaker[T, S]
+	CollectStartupMsgs         CollectStartupMsgsFunc[S]
 	CollectSubscriptionModMsgs CollectSubscriptionModMsgsFunc[S]
 }
 
-type CollectStartupMsgsFunc[S any] func(known string, spec S, ctx context.Context) (
+// CollectStartupMsgsFunc collects messages that were missed prior to startup.
+type CollectStartupMsgsFunc[S any] func(
+	ctx context.Context,
+	user model.User,
+	known string,
+	spec S,
+) (
 	[]*websocket.PreparedMessage, error,
 )
 
-type CollectSubscriptionModMsgsFunc[S any] func(addSpec S, ctx context.Context) (
+// CollectSubscriptionModMsgsFunc collects messages that are missed due to modifying a subscription.
+type CollectSubscriptionModMsgsFunc[S any] func(ctx context.Context, addSpec S) (
 	[]*websocket.PreparedMessage, error,
 )
 
+// Restart restarts underlying publishers and closes all active websocket connections.
+func (ps *PublisherSet) Restart() (errs []error) {
+	ps.Trials.CloseAllStreamers()
+	// ps.Experiments.CloseAllStreamers()
```

Comment:
question:  What happens if somebody is in-between subscriptions?  Could they dodge this close action that way?

I think explicitly tracking all streamers was a safer strategy.

Also, maybe there's an easier way, which is equally safe to tracking all streamers.  We already have a goroutine per websocket waiting for the Context to get closed, I think that same goroutine could also listen to a `bootem chan struct{}`.  You could keep one `bootem` chan per PublisherSet, and `Reset()` could lock a mutex, close the `bootem` chan, create a new chan, and unlock the mutex.  That way everybody who connected previously is guaranteed to be booted, and you don't need to add or delete from a list.

Reply:
It's either your approach with channels or implementing a solution readers-writers problem (w/ writers-preference) around reading the permissions for creating a new subscription & closing all streamers.

Reply:
Would the bootem channel approach let messages be sent with the stale permission info before seeing the that the streamer is closed?

Reply:
- streamer receives messages
- Websocket calls waitForSomething & completes
- a permission change occurs, closing all streamers
- streamer is closed, but will still send the messages